File: Python/ByRound/1762/1762_G_Unequal_Adjacent_Elements.py
# Problem: CF 1762 G - Unequal Adjacent Elements
# https://codeforces.com/contest/1762/problem/G

"""
Algorithm: Segmentation and Greedy Pairing
Time Complexity: O(n)
Space Complexity: O(n)

This solution attempts to construct a permutation of indices such that:
1. For any three consecutive elements in the permutation, the first and third are not equal.
2. For any two consecutive elements in the permutation, the corresponding array values are not equal.

The approach involves:
1. Identifying frequency of each value in the input array.
2. Checking if a valid permutation is possible (no value occurs more than ⌈n/2⌉ times).
3. Constructing segments of pairs of indices that satisfy the constraints.
4. Handling remaining indices by placing them appropriately to avoid conflicts.
5. Adjusting adjacent segments to ensure compatibility between them.
"""

import sys

input_data: list[str] = sys.stdin.read().strip().split()
data_index: int = 0


def input() -> str:
    """
    Read the next token from the pre-read input data.

    :return: A single token (string) from the input data.
    """
    global data_index
    val = input_data[data_index]
    data_index += 1
    return val


def solve_one_case() -> None:
    """
    Solve a single test case.

    This function reads an integer n, then reads an array of n elements.
    It constructs segments by pairing indices of distinct values under
    certain frequency constraints. If it is impossible to satisfy the
    constraints (i.e., one value appears more than half the time, rounded up),
    it prints "NO". Otherwise, it prints "YES" followed by the segments.
    """

    n: int = int(input())

    arr: list[int] = [0] * (n + 1)
    freq: list[int] = [0] * (n + 1)
    bucket: list[int] = [0] * (n + 1)
    visited: list[bool] = [False] * (n + 1)

    segments: list[list[int]] = []

    mx_value: int = 0

    for i in range(1, n + 1):
        arr[i] = int(input())

    for i in range(1, n + 1):
        freq[arr[i]] += 1

    for i in range(1, n + 1):
        mx_value = max(mx_value, freq[i])
        bucket[freq[i]] += 1

    if mx_value > (n + 1) // 2:
        print("NO")
        return

    def get_max() -> int:
        """
        Return the current maximum frequency for which the bucket count is non-zero.

        :return: The maximum frequency that still has a positive count.
        """
        nonlocal mx_value
        while mx_value > 0 and bucket[mx_value] == 0:
            mx_value -= 1
        return mx_value

    def reduce_freq(x: int) -> None:
        """
        Decrease the frequency of the given value x by one. The bucket data structure
        and freq array are updated accordingly.

        :param x: The value whose frequency should be reduced by one.
        """
        old_f = freq[x]
        bucket[old_f] -= 1
        freq[x] = old_f - 1
        bucket[freq[x]] += 1

    def adjust_even_pairs(vec: list[int]) -> None:
        """
        Swap elements in the list in pairs. That is, for indices (0,1), (2,3), etc.

        :param vec: The list of indices to be rearranged in-place.
        """
        for i in range(0, len(vec), 2):
            vec[i], vec[i ^ 1] = vec[i ^ 1], vec[i]

    r: int = 0
    j: int = 2
    for i in range(1, n + 1):
        if i == r + 1:
            segments.append([])

        if get_max() == ((n - r) + 1) // 2:
            break

        if not visited[i]:
            while j <= n and arr[j] == arr[i]:
                j += 1
            if j > n:
                print("NO")
                return

            visited[i] = True
            visited[j] = True

            reduce_freq(arr[i])
            reduce_freq(arr[j])

            r += 2
            segments[-1].append(j)
            segments[-1].append(i)

            j += 1

    max_position: int = max(range(1, n + 1), key=lambda x: freq[x])
    group_0: list[int] = []
    group_1: list[int] = []

    for i in range(1, n + 1):
        if not visited[i]:
            if arr[i] != max_position:
                group_1.append(i)
            else:
                group_0.append(i)

    total_left: int = len(group_0) + len(group_1)
    for i in range(total_left):
        if (i & 1) == 0:
            segments[-1].append(group_0[i >> 1])
        else:
            segments[-1].append(group_1[i >> 1])

    for idx in range(len(segments) - 1, 0, -1):
        first_in_current = segments[idx][0]
        last_in_previous = segments[idx - 1][-1]
        if arr[first_in_current] == arr[last_in_previous]:
            adjust_even_pairs(segments[idx - 1])

    print("YES")
    for seg in segments:
        for x in seg:
            print(x, end=" ")
    print()


def main() -> None:
    """
    Orchestrate the solution by reading the number of test cases,
    then solve each one by calling solve_one_case.
    """
    t: int = int(input())
    for _ in range(t):
        solve_one_case()


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/G/1762_G_Unequal_Adjacent_Elements.cpp
// Problem: CF 1762 G - Unequal Adjacent Elements
// https://codeforces.com/contest/1762/problem/G

/*
Algorithm/Technique: Greedy + Counting
Time Complexity: O(n)
Space Complexity: O(n)

This code attempts to find a permutation of indices such that:
1. For every i >= 3, p[i-2] < p[i] (i.e., no three consecutive elements in the permutation are in increasing order)
2. For every i >= 2, a[p[i-1]] != a[p[i]] (no two adjacent elements in the permutation have same value in the original array)

Approach:
- First, we count occurrences of each value in the array.
- If any value occurs more than (n+1)/2 times, it's impossible to construct such permutation.
- Otherwise, we can construct a valid permutation by placing elements in a greedy fashion:
  - Place elements with highest frequency first, ensuring no two of them are adjacent.
  - Fill the remaining positions with other elements.

The algorithm uses a greedy assignment strategy to ensure:
- Consecutive elements in the final permutation have different values in the original array.
- No three consecutive elements in the final permutation form an ascending sequence.
*/
#include <algorithm>
#include <iostream>
#include <utility>

using namespace std;
int _, n, a[1 << 20], s[1 << 20], c[1 << 20], u, A[1 << 20];
void O() {
  cin >> n, u = 0;
  for (int i = 1; i <= n; i++)
    s[i] = c[i] = 0;
  for (int i = 1; i <= n; i++)
    cin >> a[i], s[++c[a[i]]]++, u = max(u, c[a[i]]);
  if (u > (n + 1) / 2) {
    cout << "NO\n";
    return;
  }
  cout << "YES\n";
  for (int i = 1, l = 1, r = 2; i <= n; i++) {
    if (u == (n - l + 2) / 2)
      for (int j = 1; j <= n; j++)
        if (c[j] == u) {
          for (int k = i; k <= n; k++)
            if (a[k] == j)
              A[l] = k, l += 2;
            else
              A[r] = k, r += 2;
          for (int k = 1; k <= n; k++)
            cout << A[k] << ' ';
          cout << '\n';
          return;
        }
    if (a[i] != a[A[l - 1]]) {
      A[l] = i, l += 2;
      if (l > r)
        swap(l, r);
    } else
      A[r] = i, r += 2;
    s[c[a[i]]--]--, u -= !s[u];
  }
}
int main() {
  for (cin >> _; _--;)
    O();
}


// https://github.com/VaHiX/CodeForces/